File: phase-2/week-3/gin-s3-users/application.go
package main

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/aws/aws-sdk-go-v2/service/s3/types"
	"github.com/gin-gonic/gin"
)

type User struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	Email     string    `json:"email"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type CreateUserRequest struct {
	Name  string `json:"name" binding:"required"`
	Email string `json:"email" binding:"required,email"`
}

type UpdateUserRequest struct {
	Name  string `json:"name" binding:"required"`
	Email string `json:"email" binding:"required,email"`
}

type Application struct {
	S3Client *s3.Client
	Bucket   string
}

func main() {

	ctx := context.Background()

	// ---------------------------------------------------------
	// Read configuration
	// ---------------------------------------------------------

	bucket := os.Getenv("S3_BUCKET_NAME")

	if bucket == "" {
		log.Fatal("S3_BUCKET_NAME environment variable is required")
	}

	// ---------------------------------------------------------
	// Load AWS configuration
	// ---------------------------------------------------------

	cfg, err := config.LoadDefaultConfig(ctx)
	if err != nil {
		log.Fatalf("unable to load AWS configuration: %v", err)
	}

	// ---------------------------------------------------------
	// Create S3 client
	// ---------------------------------------------------------

	s3Client := s3.NewFromConfig(cfg)

	app := &Application{
		S3Client: s3Client,
		Bucket:   bucket,
	}

	// ---------------------------------------------------------
	// Gin router
	// ---------------------------------------------------------

	router := gin.Default()

	router.GET("/", func(c *gin.Context) {
		c.JSON(http.StatusOK, gin.H{
			"message": "Gin + AWS S3 User API",
		})
	})

	router.GET("/health", app.health)

	api := router.Group("/api")
	{
		api.POST("/users", app.createUser)

		api.GET("/users", app.getUsers)

		api.GET("/users/:id", app.getUser)

		api.PUT("/users/:id", app.updateUser)

		api.DELETE("/users/:id", app.deleteUser)
	}

	// ---------------------------------------------------------
	// Elastic Beanstalk provides PORT.
	// For local testing we'll use 5000.
	// ---------------------------------------------------------

	port := os.Getenv("PORT")

	if port == "" {
		port = "5000"
	}

	log.Printf("Application starting on port %s", port)
	log.Printf("S3 bucket: %s", bucket)

	if err := router.Run(":" + port); err != nil {
		log.Fatal(err)
	}
}

// ============================================================
// Health Check
// ============================================================

func (app *Application) health(c *gin.Context) {

	c.JSON(http.StatusOK, gin.H{
		"status": "UP",
	})
}

// ============================================================
// CREATE USER
//
// POST /api/users
// ============================================================

func (app *Application) createUser(c *gin.Context) {

	var request CreateUserRequest

	if err := c.ShouldBindJSON(&request); err != nil {

		c.JSON(http.StatusBadRequest, gin.H{
			"error": err.Error(),
		})

		return
	}

	id, err := generateID()
	if err != nil {

		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "failed to generate user id",
		})

		return
	}

	now := time.Now().UTC()

	user := User{
		ID:        id,
		Name:      request.Name,
		Email:     request.Email,
		CreatedAt: now,
		UpdatedAt: now,
	}

	if err := app.saveUser(c.Request.Context(), user); err != nil {

		log.Printf("failed to create user: %v", err)

		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "failed to save user to S3",
		})

		return
	}

	c.JSON(http.StatusCreated, user)
}

// ============================================================
// GET USER
//
// GET /api/users/:id
// ============================================================

func (app *Application) getUser(c *gin.Context) {

	id := c.Param("id")

	user, err := app.readUser(
		c.Request.Context(),
		id,
	)

	if err != nil {

		var noSuchKey *types.NoSuchKey

		if errors.As(err, &noSuchKey) {

			c.JSON(http.StatusNotFound, gin.H{
				"error": "user not found",
			})

			return
		}

		log.Printf("failed to get user: %v", err)

		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "failed to read user from S3",
		})

		return
	}

	c.JSON(http.StatusOK, user)
}

// ============================================================
// GET ALL USERS
//
// GET /api/users
// ============================================================

func (app *Application) getUsers(c *gin.Context) {

	ctx := c.Request.Context()

	input := &s3.ListObjectsV2Input{
		Bucket: aws.String(app.Bucket),
		Prefix: aws.String("users/"),
	}

	paginator := s3.NewListObjectsV2Paginator(
		app.S3Client,
		input,
	)

	users := make([]User, 0)

	for paginator.HasMorePages() {

		page, err := paginator.NextPage(ctx)
		if err != nil {

			log.Printf("failed listing users: %v", err)

			c.JSON(http.StatusInternalServerError, gin.H{
				"error": "failed to list S3 objects",
			})

			return
		}

		for _, object := range page.Contents {

			if object.Key == nil {
				continue
			}

			key := *object.Key

			if !strings.HasSuffix(key, ".json") {
				continue
			}

			id := strings.TrimPrefix(key, "users/")
			id = strings.TrimSuffix(id, ".json")

			user, err := app.readUser(ctx, id)

			if err != nil {

				log.Printf(
					"unable to read object %s: %v",
					key,
					err,
				)

				continue
			}

			users = append(users, user)
		}
	}

	c.JSON(http.StatusOK, users)
}

// ============================================================
// UPDATE USER
//
// PUT /api/users/:id
// ============================================================

func (app *Application) updateUser(c *gin.Context) {

	id := c.Param("id")

	var request UpdateUserRequest

	if err := c.ShouldBindJSON(&request); err != nil {

		c.JSON(http.StatusBadRequest, gin.H{
			"error": err.Error(),
		})

		return
	}

	// First verify that the user exists.

	existingUser, err := app.readUser(
		c.Request.Context(),
		id,
	)

	if err != nil {

		var noSuchKey *types.NoSuchKey

		if errors.As(err, &noSuchKey) {

			c.JSON(http.StatusNotFound, gin.H{
				"error": "user not found",
			})

			return
		}

		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "failed to read user",
		})

		return
	}

	existingUser.Name = request.Name
	existingUser.Email = request.Email
	existingUser.UpdatedAt = time.Now().UTC()

	if err := app.saveUser(
		c.Request.Context(),
		existingUser,
	); err != nil {

		log.Printf("failed updating user: %v", err)

		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "failed to update user",
		})

		return
	}

	c.JSON(http.StatusOK, existingUser)
}

// ============================================================
// DELETE USER
//
// DELETE /api/users/:id
// ============================================================

func (app *Application) deleteUser(c *gin.Context) {

	id := c.Param("id")

	// Verify user exists before deleting.

	_, err := app.readUser(
		c.Request.Context(),
		id,
	)

	if err != nil {

		var noSuchKey *types.NoSuchKey

		if errors.As(err, &noSuchKey) {

			c.JSON(http.StatusNotFound, gin.H{
				"error": "user not found",
			})

			return
		}

		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "failed to check user",
		})

		return
	}

	key := userKey(id)

	_, err = app.S3Client.DeleteObject(
		c.Request.Context(),
		&s3.DeleteObjectInput{
			Bucket: aws.String(app.Bucket),
			Key:    aws.String(key),
		},
	)

	if err != nil {

		log.Printf("failed deleting user: %v", err)

		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "failed to delete user",
		})

		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "user deleted successfully",
		"id":      id,
	})
}

// ============================================================
// Save user to S3
// ============================================================

func (app *Application) saveUser(
	ctx context.Context,
	user User,
) error {

	data, err := json.MarshalIndent(
		user,
		"",
		"  ",
	)

	if err != nil {
		return err
	}

	key := userKey(user.ID)

	_, err = app.S3Client.PutObject(
		ctx,
		&s3.PutObjectInput{
			Bucket:      aws.String(app.Bucket),
			Key:         aws.String(key),
			Body:        bytes.NewReader(data),
			ContentType: aws.String("application/json"),
		},
	)

	return err
}

// ============================================================
// Read user from S3
// ============================================================

func (app *Application) readUser(
	ctx context.Context,
	id string,
) (User, error) {

	var user User

	key := userKey(id)

	result, err := app.S3Client.GetObject(
		ctx,
		&s3.GetObjectInput{
			Bucket: aws.String(app.Bucket),
			Key:    aws.String(key),
		},
	)

	if err != nil {
		return user, err
	}

	defer result.Body.Close()

	data, err := io.ReadAll(result.Body)

	if err != nil {
		return user, err
	}

	if err := json.Unmarshal(data, &user); err != nil {
		return user, err
	}

	return user, nil
}

// ============================================================
// Build S3 key
// ============================================================

func userKey(id string) string {

	return fmt.Sprintf(
		"users/%s.json",
		id,
	)
}

// ============================================================
// Generate random ID
// ============================================================

func generateID() (string, error) {

	bytes := make([]byte, 8)

	_, err := rand.Read(bytes)
	if err != nil {
		return "", err
	}

	return hex.EncodeToString(bytes), nil
}

// demo-users-2026
// GinUsersS3Polic

File: phase-2/week-3/python-s3-beanstalk/application.py
import json
import os
import uuid
from datetime import datetime, timezone

import boto3
from botocore.exceptions import ClientError
from flask import Flask, jsonify, request


# ---------------------------------------------------------
# Elastic Beanstalk expects a WSGI application
# ---------------------------------------------------------

application = Flask(__name__)

# Optional alias for local development
app = application


# ---------------------------------------------------------
# Configuration
# ---------------------------------------------------------

S3_BUCKET_NAME = os.getenv("S3_BUCKET_NAME", "demo-users-2026")

s3_client = boto3.client("s3")


# ---------------------------------------------------------
# Utility functions
# ---------------------------------------------------------

def generate_id():
    return uuid.uuid4().hex[:16]


def user_key(user_id):
    return f"users/{user_id}.json"


def save_user(user):
    key = user_key(user["id"])

    s3_client.put_object(
        Bucket=S3_BUCKET_NAME,
        Key=key,
        Body=json.dumps(user, indent=2).encode("utf-8"),
        ContentType="application/json"
    )


def read_user(user_id):
    key = user_key(user_id)

    try:
        response = s3_client.get_object(
            Bucket=S3_BUCKET_NAME,
            Key=key
        )

        content = response["Body"].read()

        return json.loads(content)

    except ClientError as e:

        error_code = e.response["Error"]["Code"]

        if error_code in ("NoSuchKey", "404"):
            return None

        raise


# ---------------------------------------------------------
# Root endpoint
# ---------------------------------------------------------

@application.route("/", methods=["GET"])
def root():
    return jsonify({
        "message": "Flask + AWS S3 User API"
    })


# ---------------------------------------------------------
# Health endpoint
# ---------------------------------------------------------

@application.route("/health", methods=["GET"])
def health():
    return jsonify({
        "status": "UP"
    })


# ---------------------------------------------------------
# CREATE USER
#
# POST /api/users
# ---------------------------------------------------------

@application.route("/api/users", methods=["POST"])
def create_user():

    data = request.get_json()

    if not data:
        return jsonify({
            "error": "JSON body is required"
        }), 400

    name = data.get("name")
    email = data.get("email")

    if not name:
        return jsonify({
            "error": "name is required"
        }), 400

    if not email:
        return jsonify({
            "error": "email is required"
        }), 400

    now = datetime.now(timezone.utc).isoformat()

    user = {
        "id": generate_id(),
        "name": name,
        "email": email,
        "created_at": now,
        "updated_at": now
    }

    try:

        save_user(user)

        return jsonify(user), 201

    except ClientError as e:

        application.logger.exception(
            "Failed to save user to S3"
        )

        return jsonify({
            "error": "failed to save user to S3",
            "details": str(e)
        }), 500


# ---------------------------------------------------------
# GET ALL USERS
#
# GET /api/users
# ---------------------------------------------------------

@application.route("/api/users", methods=["GET"])
def get_users():

    users = []

    try:

        paginator = s3_client.get_paginator("list_objects_v2")

        pages = paginator.paginate(
            Bucket=S3_BUCKET_NAME,
            Prefix="users/"
        )

        for page in pages:

            for obj in page.get("Contents", []):

                key = obj["Key"]

                if not key.endswith(".json"):
                    continue

                user_id = key.removeprefix("users/")
                user_id = user_id.removesuffix(".json")

                user = read_user(user_id)

                if user:
                    users.append(user)

        return jsonify(users)

    except ClientError as e:

        application.logger.exception(
            "Failed to list users"
        )

        return jsonify({
            "error": "failed to list users",
            "details": str(e)
        }), 500


# ---------------------------------------------------------
# GET ONE USER
#
# GET /api/users/<id>
# ---------------------------------------------------------

@application.route("/api/users/<user_id>", methods=["GET"])
def get_user(user_id):

    try:

        user = read_user(user_id)

        if user is None:

            return jsonify({
                "error": "user not found"
            }), 404

        return jsonify(user)

    except ClientError as e:

        application.logger.exception(
            "Failed to read user"
        )

        return jsonify({
            "error": "failed to read user from S3",
            "details": str(e)
        }), 500


# ---------------------------------------------------------
# UPDATE USER
#
# PUT /api/users/<id>
# ---------------------------------------------------------

@application.route("/api/users/<user_id>", methods=["PUT"])
def update_user(user_id):

    data = request.get_json()

    if not data:

        return jsonify({
            "error": "JSON body is required"
        }), 400

    try:

        existing_user = read_user(user_id)

        if existing_user is None:

            return jsonify({
                "error": "user not found"
            }), 404

        name = data.get("name")
        email = data.get("email")

        if not name:

            return jsonify({
                "error": "name is required"
            }), 400

        if not email:

            return jsonify({
                "error": "email is required"
            }), 400

        existing_user["name"] = name
        existing_user["email"] = email
        existing_user["updated_at"] = (
            datetime.now(timezone.utc).isoformat()
        )

        save_user(existing_user)

        return jsonify(existing_user)

    except ClientError as e:

        application.logger.exception(
            "Failed to update user"
        )

        return jsonify({
            "error": "failed to update user",
            "details": str(e)
        }), 500


# ---------------------------------------------------------
# DELETE USER
#
# DELETE /api/users/<id>
# ---------------------------------------------------------

@application.route(
    "/api/users/<user_id>",
    methods=["DELETE"]
)
def delete_user(user_id):

    try:

        existing_user = read_user(user_id)

        if existing_user is None:

            return jsonify({
                "error": "user not found"
            }), 404

        key = user_key(user_id)

        s3_client.delete_object(
            Bucket=S3_BUCKET_NAME,
            Key=key
        )

        return jsonify({
            "message": "user deleted successfully",
            "id": user_id
        })

    except ClientError as e:

        application.logger.exception(
            "Failed to delete user"
        )

        return jsonify({
            "error": "failed to delete user",
            "details": str(e)
        }), 500


# ---------------------------------------------------------
# Local development only
# ---------------------------------------------------------

if __name__ == "__main__":

    port = int(os.getenv("PORT", "5001"))

    application.run(
        host="0.0.0.0",
        port=port,
        debug=False
    )